--- ipset-6.6/configure.ac.orig	2011-05-24 10:28:55.000000000 +0200
+++ ipset-6.6/configure.ac	2011-05-26 20:47:06.586245363 +0200
@@ -39,23 +39,6 @@
 else
 	ksourcedir="/lib/modules/$(uname -r)/source"
 fi
-if test ! -e "$ksourcedir/include/linux/netfilter/nfnetlink.h"
-then
-	AC_MSG_ERROR([Invalid kernel source directory $ksourcedir])
-fi
-
-if test ! -e "$kbuilddir/.config"
-then
-	AC_MSG_ERROR([The kernel build directory $kbuilddir is not configured])
-fi
-
-AC_PROG_GREP
-
-if test "X`$GREP 'NFNL_SUBSYS_IPSET' $ksourcedir/include/linux/netfilter/nfnetlink.h`" = "X"
-then
-	AC_MSG_ERROR([The kernel source directory $ksourcedir is not patched with netlink.patch to support ipset])
-fi
-
 dnl Maximal number of sets supported by the kernel, default 256
 AC_ARG_WITH([maxsets],
 	    AS_HELP_STRING([--with-maxsets=256],
