--- ipset-6.20/configure.ac.orig	2013-10-02 22:53:23.000000000 +0200
+++ ipset-6.20/configure.ac	2013-10-24 17:52:20.931877137 +0200
@@ -167,6 +167,7 @@
 dnl Checks for functions
 AC_CHECK_FUNCS(gethostbyname2)
 
+if test "$BUILDKMOD" == "yes"; then
 dnl Check kernel incompatibilities... Ugly like hell
 AC_MSG_CHECKING([kernel source for struct xt_action_param])
 if test -f $ksourcedir/include/linux/netfilter/x_tables.h && \
@@ -309,6 +310,7 @@
 	AC_MSG_RESULT(no)
 	AC_MSG_ERROR([Netns support is required in the Linux kernel tree])
 fi
+fi
 
 dnl Checks for compiler characteristics.
 dnl Check extra warning flags except
