--- ipset-7.1/configure.ac~	2018-12-10 18:35:31.000000000 +0100
+++ ipset-7.1/configure.ac	2019-03-10 18:03:26.965123071 +0100
@@ -581,7 +581,7 @@
 
 AC_MSG_CHECKING([kernel source for netlink extended ACK])
 if test -f $ksourcedir/include/linux/netlink.h && \
-   $AWK '/^extern void netlink_ack\(/,/\)/' $ksourcedir/include/linux/netlink.h | $GREP -q 'const struct netlink_ext_ack'; then
+   $AWK '/^void netlink_ack\(/,/\)/' $ksourcedir/include/linux/netlink.h | $GREP -q 'const struct netlink_ext_ack'; then
 	AC_MSG_RESULT(yes)
 	AC_SUBST(HAVE_NETLINK_EXTENDED_ACK, define)
 else
