--- ipset-7.15/configure.ac~	2021-08-04 08:18:14.000000000 +0200
+++ ipset-7.15/configure.ac	2022-01-20 22:11:42.308828836 +0100
@@ -450,7 +450,7 @@
 
 AC_MSG_CHECKING([kernel source for kvfree])
 if test -f $ksourcedir/include/linux/mm.h && \
-   $GREP -q 'kvfree' $ksourcedir/include/linux/mm.h; then
+   $GREP -q 'kvfree' $ksourcedir/include/linux/mm.h $ksourcedir/include/linux/slab.h; then
 	AC_MSG_RESULT(yes)
 	AC_SUBST(HAVE_KVFREE, define)
 else
--- ipset-7.15/configure.ac~	2022-01-20 22:12:29.000000000 +0100
+++ ipset-7.15/configure.ac	2022-01-20 22:13:17.434329590 +0100
@@ -839,7 +839,7 @@
 
 AC_MSG_CHECKING([kernel source for kvzalloc() in mm.h])
 if test -f $ksourcedir/include/linux/mm.h && \
-   $GREP -q 'static inline void \*kvzalloc(' $ksourcedir/include/linux/mm.h; then
+   $GREP -q ' void \*kvzalloc(' $ksourcedir/include/linux/mm.h $ksourcedir/include/linux/slab.h; then
 	AC_MSG_RESULT(yes)
 	AC_SUBST(HAVE_KVZALLOC, define)
 else
