--- ipset/ipset-6.24/kernel/net/netfilter/ipset/ip_set_list_set.c~	2014-11-24 21:46:45.000000000 +0100
+++ ipset/ipset-6.24/kernel/net/netfilter/ipset/ip_set_list_set.c	2015-01-28 19:09:21.324834608 +0100
@@ -280,7 +280,11 @@
 	if (d->before == 0) {
 		/* Append  */
 		n = list_empty(&map->members) ? NULL :
+#if LINUX_VERSION_CODE > KERNEL_VERSION(3,12,0)
 		    list_last_entry(&map->members, struct set_elem, list);
+#else
+		    list_entry((&map->members)->prev, struct set_elem, list)
+#endif
 	} else if (d->before > 0) {
 		/* Insert after next element */
 		if (!list_is_last(&next->list, &map->members))
